In [36]:
# Câu 1
#  1. Tạo dữ liệu 2 bảng
import pandas as pd
import sqlite3

# Kết nối SQLite trong bộ nhớ
conn = sqlite3.connect(":memory:")
cursor = conn.cursor()

# Tạo bảng student
cursor.execute("""
CREATE TABLE student (
    student_id INTEGER,
    name TEXT,
    class TEXT,
    course_id INTEGER,
    score REAL
)
""")

# Tạo bảng course
cursor.execute("""
CREATE TABLE course (
    id INTEGER,
    course_name TEXT
)
""")

# Chèn dữ liệu vào bảng student
students = [
    (1, "Nguyen Minh Hoang", "May Tinh", 12, 6.7),
    (2, "Tran Thi Lan", "Kinh Te", 34, 9.2),
    (3, "Pham Van Nam", "Toan Tin", None, 7.9),
    (4, "Le Thanh Huyen", "Toan Tin", 20, 7.2),
    (5, "Vu Quoc Anh", "May Tinh", 24, 8.0),
    (6, "Dang Thuy Linh", "May Tinh", 24, 5.5),
    (7, "Bui Tien Dung", "Kinh Te", 34, 9.2),
    (8, "Ho Ngoc Mai", "Toan Tin", 20, 8.8),
    (9, "Duong Huu Phuc", "Kinh Te", None, 7.2),
    (10, "Cao Thi Hanh", "May Tinh", None, 7.0)
]

cursor.executemany("INSERT INTO student VALUES (?, ?, ?, ?, ?)", students)

# Chèn dữ liệu vào bảng course
courses = [
    (12, "Giai tich"),
    (34, "Thong ke"),
    (26, "Tin hoc")
]

cursor.executemany("INSERT INTO course VALUES (?, ?)", courses)

conn.commit()

In [7]:
# Hiển thị 2 bảng gốc
print("Bảng student:")
display(pd.read_sql("SELECT * FROM student", conn))

print("Bảng course:")
display(pd.read_sql("SELECT * FROM course", conn))

Bảng student:


,student_id,name,class,course_id,score
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7
1,2,Tran Thi Lan,Kinh Te,34.0,9.2
2,3,Pham Van Nam,Toan Tin,NaN,7.9
3,4,Le Thanh Huyen,Toan Tin,20.0,7.2
4,5,Vu Quoc Anh,May Tinh,24.0,8.0
5,6,Dang Thuy Linh,May Tinh,24.0,5.5
6,7,Bui Tien Dung,Kinh Te,34.0,9.2
7,8,Ho Ngoc Mai,Toan Tin,20.0,8.8
8,9,Duong Huu Phuc,Kinh Te,NaN,7.2
9,10,Cao Thi Hanh,May Tinh,NaN,7.0


Bảng course:


,id,course_name
0,12,Giai tich
1,34,Thong ke
2,26,Tin hoc


In [9]:
# Tích Descartes (CROSS JOIN)
query_cross = """
SELECT *
FROM student
CROSS JOIN course
"""
df_cross = pd.read_sql(query_cross, conn)
display(df_cross)
print("Số dòng tích Descartes:", len(df_cross))


,student_id,name,class,course_id,score,id,course_name
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7,12,Giai tich
1,1,Nguyen Minh Hoang,May Tinh,12.0,6.7,34,Thong ke
2,1,Nguyen Minh Hoang,May Tinh,12.0,6.7,26,Tin hoc
3,2,Tran Thi Lan,Kinh Te,34.0,9.2,12,Giai tich
4,2,Tran Thi Lan,Kinh Te,34.0,9.2,34,Thong ke
5,2,Tran Thi Lan,Kinh Te,34.0,9.2,26,Tin hoc
6,3,Pham Van Nam,Toan Tin,NaN,7.9,12,Giai tich
7,3,Pham Van Nam,Toan Tin,NaN,7.9,34,Thong ke
8,3,Pham Van Nam,Toan Tin,NaN,7.9,26,Tin hoc
9,4,Le Thanh Huyen,Toan Tin,20.0,7.2,12,Giai tich


Số dòng tích Descartes: 30


In [10]:
# INNER JOIN
query_inner = """
SELECT s.student_id, s.name, s.class, s.course_id, s.score, c.course_name
FROM student s
INNER JOIN course c
ON s.course_id = c.id
"""
df_inner = pd.read_sql(query_inner, conn)
display(df_inner)

,student_id,name,class,course_id,score,course_name
0,1,Nguyen Minh Hoang,May Tinh,12,6.7,Giai tich
1,2,Tran Thi Lan,Kinh Te,34,9.2,Thong ke
2,7,Bui Tien Dung,Kinh Te,34,9.2,Thong ke


In [11]:
# LEFT JOIN
query_left = """
SELECT s.student_id, s.name, s.class, s.course_id, s.score, c.course_name
FROM student s
LEFT JOIN course c
ON s.course_id = c.id
"""
df_left = pd.read_sql(query_left, conn)
display(df_left)

,student_id,name,class,course_id,score,course_name
0,1,Nguyen Minh Hoang,May Tinh,12.0,6.7,Giai tich
1,2,Tran Thi Lan,Kinh Te,34.0,9.2,Thong ke
2,3,Pham Van Nam,Toan Tin,NaN,7.9,NaN
3,4,Le Thanh Huyen,Toan Tin,20.0,7.2,NaN
4,5,Vu Quoc Anh,May Tinh,24.0,8.0,NaN
5,6,Dang Thuy Linh,May Tinh,24.0,5.5,NaN
6,7,Bui Tien Dung,Kinh Te,34.0,9.2,Thong ke
7,8,Ho Ngoc Mai,Toan Tin,20.0,8.8,NaN
8,9,Duong Huu Phuc,Kinh Te,NaN,7.2,NaN
9,10,Cao Thi Hanh,May Tinh,NaN,7.0,NaN


In [12]:
# RIGHT JOIN
query_right = """
SELECT s.student_id, s.name, s.class, s.course_id, s.score, c.course_name
FROM course c
LEFT JOIN student s
ON s.course_id = c.id
"""
df_right = pd.read_sql(query_right, conn)
display(df_right)

,student_id,name,class,course_id,score,course_name
0,1.0,Nguyen Minh Hoang,May Tinh,12.0,6.7,Giai tich
1,2.0,Tran Thi Lan,Kinh Te,34.0,9.2,Thong ke
2,7.0,Bui Tien Dung,Kinh Te,34.0,9.2,Thong ke
3,NaN,NaN,NaN,NaN,NaN,Tin hoc


In [13]:
# FULL OUTER JOIN
query_full = """
SELECT s.student_id, s.name, s.class, s.course_id, s.score, c.id, c.course_name
FROM student s
LEFT JOIN course c
ON s.course_id = c.id

UNION

SELECT s.student_id, s.name, s.class, s.course_id, s.score, c.id, c.course_name
FROM course c
LEFT JOIN student s
ON s.course_id = c.id
"""
df_full = pd.read_sql(query_full, conn)
display(df_full)

,student_id,name,class,course_id,score,id,course_name
0,NaN,NaN,NaN,NaN,NaN,26.0,Tin hoc
1,1.0,Nguyen Minh Hoang,May Tinh,12.0,6.7,12.0,Giai tich
2,2.0,Tran Thi Lan,Kinh Te,34.0,9.2,34.0,Thong ke
3,3.0,Pham Van Nam,Toan Tin,NaN,7.9,NaN,NaN
4,4.0,Le Thanh Huyen,Toan Tin,20.0,7.2,NaN,NaN
5,5.0,Vu Quoc Anh,May Tinh,24.0,8.0,NaN,NaN
6,6.0,Dang Thuy Linh,May Tinh,24.0,5.5,NaN,NaN
7,7.0,Bui Tien Dung,Kinh Te,34.0,9.2,34.0,Thong ke
8,8.0,Ho Ngoc Mai,Toan Tin,20.0,8.8,NaN,NaN
9,9.0,Duong Huu Phuc,Kinh Te,NaN,7.2,NaN,NaN


Kết luận câu 1:

* Đã thực hiện kết nối hai bảng student và course bằng các cách: tích Descartes, INNER JOIN, LEFT JOIN, RIGHT JOIN mô phỏng và FULL OUTER JOIN mô phỏng.
* Kết quả CROSS JOIN cho ra tất cả các tổ hợp giữa dữ liệu của hai bảng.
* INNER JOIN chỉ giữ lại các bản ghi có course_id của sinh viên trùng với id trong bảng course.
* LEFT JOIN giữ toàn bộ dữ liệu của bảng student, kể cả những sinh viên chưa có course_id hoặc có course_id không tồn tại trong bảng course.
* RIGHT JOIN mô phỏng giữ toàn bộ dữ liệu của bảng course, nên có thể thấy môn Tin hoc vẫn xuất hiện dù chưa có sinh viên nào khớp.
* FULL OUTER JOIN mô phỏng cho thấy đầy đủ dữ liệu từ cả hai bảng; các giá trị NaN hoặc None xuất hiện tại những vị trí không có bản ghi tương ứng giữa hai bảng.
* Như vậy, câu 1 đã cho thấy sự khác nhau giữa từng kiểu nối bảng và giúp xác định các bản ghi khớp hoặc không khớp giữa student và course

In [ ]:
# Câu 2
# 1. Cập nhật các course_id còn thiếu
# Cập nhật các course_id còn thiếu bằng các giá trị có trong bảng course
cursor.execute("UPDATE student SET course_id = 26 WHERE student_id = 3")
cursor.execute("UPDATE student SET course_id = 12 WHERE student_id = 9")
cursor.execute("UPDATE student SET course_id = 34 WHERE student_id = 10")
conn.commit()

print("Sau khi cập nhật course_id bị thiếu:")
display(pd.read_sql("SELECT * FROM student", conn))

Sau khi cập nhật course_id bị thiếu:


,student_id,name,class,course_id,score
0,1,Nguyen Minh Hoang,May Tinh,12,6.7
1,2,Tran Thi Lan,Kinh Te,34,9.2
2,3,Pham Van Nam,Toan Tin,26,7.9
3,4,Le Thanh Huyen,Toan Tin,20,7.2
4,5,Vu Quoc Anh,May Tinh,24,8.0
5,6,Dang Thuy Linh,May Tinh,24,5.5
6,7,Bui Tien Dung,Kinh Te,34,9.2
7,8,Ho Ngoc Mai,Toan Tin,20,8.8
8,9,Duong Huu Phuc,Kinh Te,12,7.2
9,10,Cao Thi Hanh,May Tinh,34,7.0


In [ ]:
# 2. Loại bỏ các bản ghi có course_id không tồn tại trong bảng course
cursor.execute("""
DELETE FROM student
WHERE course_id NOT IN (SELECT id FROM course)
""")
conn.commit()

print("Sau khi xóa các bản ghi có course_id không tồn tại trong course:")
display(pd.read_sql("SELECT * FROM student", conn))

Sau khi xóa các bản ghi có course_id không tồn tại trong course:


,student_id,name,class,course_id,score
0,1,Nguyen Minh Hoang,May Tinh,12,6.7
1,2,Tran Thi Lan,Kinh Te,34,9.2
2,3,Pham Van Nam,Toan Tin,26,7.9
3,7,Bui Tien Dung,Kinh Te,34,9.2
4,9,Duong Huu Phuc,Kinh Te,12,7.2
5,10,Cao Thi Hanh,May Tinh,34,7.0


In [ ]:
# 3. Tổng số sinh viên và điểm trung bình theo từng lớp
query_class = """
SELECT 
    class,
    COUNT(*) AS tong_so_sinh_vien,
    ROUND(AVG(score), 2) AS diem_trung_binh
FROM student
GROUP BY class
ORDER BY class
"""
df_class = pd.read_sql(query_class, conn)
print("Tổng số sinh viên và điểm trung bình theo từng lớp:")
display(df_class)

Tổng số sinh viên và điểm trung bình theo từng lớp:


,class,tong_so_sinh_vien,diem_trung_binh
0,Kinh Te,3,8.53
1,May Tinh,2,6.85
2,Toan Tin,1,7.90


In [ ]:
# 4. Tổng số sinh viên và điểm trung bình theo từng môn học
query_course = """
SELECT 
    c.id AS course_id,
    c.course_name,
    COUNT(s.student_id) AS tong_so_sinh_vien,
    ROUND(AVG(s.score), 2) AS diem_trung_binh
FROM course c
LEFT JOIN student s
ON s.course_id = c.id
GROUP BY c.id, c.course_name
ORDER BY c.id
"""
df_course = pd.read_sql(query_course, conn)
print("Tổng số sinh viên và điểm trung bình theo từng môn học:")
display(df_course)

Tổng số sinh viên và điểm trung bình theo từng môn học:


,course_id,course_name,tong_so_sinh_vien,diem_trung_binh
0,12,Giai tich,2,6.95
1,26,Tin hoc,1,7.90
2,34,Thong ke,3,8.47


In [ ]:
# 5. Phân loại từng môn học theo điểm trung bình
query_rank_course = """
SELECT 
    c.id AS course_id,
    c.course_name,
    COUNT(s.student_id) AS tong_so_sinh_vien,
    ROUND(AVG(s.score), 2) AS diem_trung_binh,
    CASE
        WHEN AVG(s.score) >= 9.0 THEN 'Xuat sac'
        WHEN AVG(s.score) >= 6.0 AND AVG(s.score) <= 8.9 THEN 'Tot'
        ELSE 'Kem'
    END AS xep_loai
FROM course c
LEFT JOIN student s
ON s.course_id = c.id
GROUP BY c.id, c.course_name
ORDER BY c.id
"""
df_rank_course = pd.read_sql(query_rank_course, conn)
print("Phân loại môn học theo điểm trung bình:")
display(df_rank_course)

Phân loại môn học theo điểm trung bình:


,course_id,course_name,tong_so_sinh_vien,diem_trung_binh,xep_loai
0,12,Giai tich,2,6.95,Tot
1,26,Tin hoc,1,7.90,Tot
2,34,Thong ke,3,8.47,Tot


Kết luận câu 2
* Các giá trị course_id còn thiếu trong bảng student đã được cập nhật bằng những mã môn có tồn tại trong bảng course. Sau đó, các bản ghi có course_id không hợp lệ đã được xóa khỏi bảng student.
* Sau khi làm sạch dữ liệu, bảng student còn 6 sinh viên hợp lệ để tiếp tục thống kê và xếp hạng.
* Thống kê theo lớp cho thấy: lớp Kinh Te có 3 sinh viên, điểm trung bình 8.53; lớp May Tinh có 2 sinh viên, điểm trung bình 6.85; lớp Toan Tin có 1 sinh viên, điểm trung bình 7.90. Trong đó lớp Kinh Te có điểm trung bình cao nhất.
* Thống kê theo môn cho thấy: môn Giai tich có 2 sinh viên, điểm trung bình 6.95; môn Tin hoc có 1 sinh viên, điểm trung bình 7.90; môn Thong ke có 3 sinh viên, điểm trung bình 8.47. Môn Thong ke có số sinh viên nhiều nhất và điểm trung bình cao nhất.
* Dựa trên điểm trung bình, cả 3 môn học đều được xếp loại Tốt theo tiêu chí của đề bài

In [ ]:
# Câu 3
# 1. Xếp hạng sinh viên theo điểm số toàn bộ
query_rank_all = """
SELECT
    student_id,
    name,
    class,
    course_id,
    score,
    RANK() OVER (ORDER BY score DESC) AS ranking
FROM student
"""
df_rank_all = pd.read_sql(query_rank_all, conn)
print("Xếp hạng sinh viên theo điểm số toàn bộ:")
display(df_rank_all)

Xếp hạng sinh viên theo điểm số toàn bộ:


,student_id,name,class,course_id,score,ranking
0,2,Tran Thi Lan,Kinh Te,34,9.2,1
1,7,Bui Tien Dung,Kinh Te,34,9.2,1
2,3,Pham Van Nam,Toan Tin,26,7.9,3
3,9,Duong Huu Phuc,Kinh Te,12,7.2,4
4,10,Cao Thi Hanh,May Tinh,34,7.0,5
5,1,Nguyen Minh Hoang,May Tinh,12,6.7,6


In [ ]:
# 2. Top 3 sinh viên điểm cao nhất toàn bộ
query_top3_all = """
SELECT *
FROM (
    SELECT
        student_id,
        name,
        class,
        course_id,
        score,
        RANK() OVER (ORDER BY score DESC) AS ranking
    FROM student
)
WHERE ranking <= 3
"""
df_top3_all = pd.read_sql(query_top3_all, conn)
print("Top 3 sinh viên điểm cao nhất toàn bộ:")
display(df_top3_all)

Top 3 sinh viên điểm cao nhất toàn bộ:


,student_id,name,class,course_id,score,ranking
0,2,Tran Thi Lan,Kinh Te,34,9.2,1
1,7,Bui Tien Dung,Kinh Te,34,9.2,1
2,3,Pham Van Nam,Toan Tin,26,7.9,3


In [ ]:
# 3. Top 3 sinh viên điểm thấp nhất toàn bộ
query_bottom3_all = """
SELECT *
FROM (
    SELECT
        student_id,
        name,
        class,
        course_id,
        score,
        RANK() OVER (ORDER BY score ASC) AS ranking
    FROM student
)
WHERE ranking <= 3
"""
df_bottom3_all = pd.read_sql(query_bottom3_all, conn)
print("Top 3 sinh viên điểm thấp nhất toàn bộ:")
display(df_bottom3_all)

Top 3 sinh viên điểm thấp nhất toàn bộ:


,student_id,name,class,course_id,score,ranking
0,1,Nguyen Minh Hoang,May Tinh,12,6.7,1
1,10,Cao Thi Hanh,May Tinh,34,7.0,2
2,9,Duong Huu Phuc,Kinh Te,12,7.2,3


In [ ]:
# 4. Xếp hạng sinh viên theo điểm số trong từng lớp
query_rank_class = """
SELECT
    student_id,
    name,
    class,
    course_id,
    score,
    RANK() OVER (PARTITION BY class ORDER BY score DESC) AS ranking_in_class
FROM student
ORDER BY class, ranking_in_class
"""
df_rank_class = pd.read_sql(query_rank_class, conn)
print("Xếp hạng sinh viên theo từng lớp:")
display(df_rank_class)

Xếp hạng sinh viên theo từng lớp:


,student_id,name,class,course_id,score,ranking_in_class
0,2,Tran Thi Lan,Kinh Te,34,9.2,1
1,7,Bui Tien Dung,Kinh Te,34,9.2,1
2,9,Duong Huu Phuc,Kinh Te,12,7.2,3
3,10,Cao Thi Hanh,May Tinh,34,7.0,1
4,1,Nguyen Minh Hoang,May Tinh,12,6.7,2
5,3,Pham Van Nam,Toan Tin,26,7.9,1


In [ ]:
# 5. Top 3 cao nhất theo từng lớp
query_top3_class = """
SELECT *
FROM (
    SELECT
        student_id,
        name,
        class,
        course_id,
        score,
        RANK() OVER (PARTITION BY class ORDER BY score DESC) AS ranking_in_class
    FROM student
)
WHERE ranking_in_class <= 3
ORDER BY class, ranking_in_class
"""
df_top3_class = pd.read_sql(query_top3_class, conn)
print("Top 3 cao nhất theo từng lớp:")
display(df_top3_class)

Top 3 cao nhất theo từng lớp:


,student_id,name,class,course_id,score,ranking_in_class
0,2,Tran Thi Lan,Kinh Te,34,9.2,1
1,7,Bui Tien Dung,Kinh Te,34,9.2,1
2,9,Duong Huu Phuc,Kinh Te,12,7.2,3
3,10,Cao Thi Hanh,May Tinh,34,7.0,1
4,1,Nguyen Minh Hoang,May Tinh,12,6.7,2
5,3,Pham Van Nam,Toan Tin,26,7.9,1


In [ ]:
# 6. Top 3 thấp nhất theo từng lớp
query_bottom3_class = """
SELECT *
FROM (
    SELECT
        student_id,
        name,
        class,
        course_id,
        score,
        RANK() OVER (PARTITION BY class ORDER BY score ASC) AS ranking_in_class
    FROM student
)
WHERE ranking_in_class <= 3
ORDER BY class, ranking_in_class
"""
df_bottom3_class = pd.read_sql(query_bottom3_class, conn)
print("Top 3 thấp nhất theo từng lớp:")
display(df_bottom3_class)

Top 3 thấp nhất theo từng lớp:


,student_id,name,class,course_id,score,ranking_in_class
0,9,Duong Huu Phuc,Kinh Te,12,7.2,1
1,2,Tran Thi Lan,Kinh Te,34,9.2,2
2,7,Bui Tien Dung,Kinh Te,34,9.2,2
3,1,Nguyen Minh Hoang,May Tinh,12,6.7,1
4,10,Cao Thi Hanh,May Tinh,34,7.0,2
5,3,Pham Van Nam,Toan Tin,26,7.9,1


In [ ]:
# 7. Xếp hạng sinh viên theo điểm số trong từng mã môn học
query_rank_course = """
SELECT
    student_id,
    name,
    class,
    course_id,
    score,
    RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS ranking_in_course
FROM student
ORDER BY course_id, ranking_in_course
"""
df_rank_course = pd.read_sql(query_rank_course, conn)
print("Xếp hạng sinh viên theo từng mã môn học:")
display(df_rank_course)

Xếp hạng sinh viên theo từng mã môn học:


,student_id,name,class,course_id,score,ranking_in_course
0,9,Duong Huu Phuc,Kinh Te,12,7.2,1
1,1,Nguyen Minh Hoang,May Tinh,12,6.7,2
2,3,Pham Van Nam,Toan Tin,26,7.9,1
3,2,Tran Thi Lan,Kinh Te,34,9.2,1
4,7,Bui Tien Dung,Kinh Te,34,9.2,1
5,10,Cao Thi Hanh,May Tinh,34,7.0,3


In [ ]:
# 8. Top 3 cao nhất theo từng mã môn học
query_top3_course = """
SELECT *
FROM (
    SELECT
        student_id,
        name,
        class,
        course_id,
        score,
        RANK() OVER (PARTITION BY course_id ORDER BY score DESC) AS ranking_in_course
    FROM student
)
WHERE ranking_in_course <= 3
ORDER BY course_id, ranking_in_course
"""
df_top3_course = pd.read_sql(query_top3_course, conn)
print("Top 3 cao nhất theo từng mã môn học:")
display(df_top3_course)

Top 3 cao nhất theo từng mã môn học:


,student_id,name,class,course_id,score,ranking_in_course
0,9,Duong Huu Phuc,Kinh Te,12,7.2,1
1,1,Nguyen Minh Hoang,May Tinh,12,6.7,2
2,3,Pham Van Nam,Toan Tin,26,7.9,1
3,2,Tran Thi Lan,Kinh Te,34,9.2,1
4,7,Bui Tien Dung,Kinh Te,34,9.2,1
5,10,Cao Thi Hanh,May Tinh,34,7.0,3


In [ ]:
# 9. Top 3 thấp nhất theo từng mã môn học
query_bottom3_course = """
SELECT *
FROM (
    SELECT
        student_id,
        name,
        class,
        course_id,
        score,
        RANK() OVER (PARTITION BY course_id ORDER BY score ASC) AS ranking_in_course
    FROM student
)
WHERE ranking_in_course <= 3
ORDER BY course_id, ranking_in_course
"""
df_bottom3_course = pd.read_sql(query_bottom3_course, conn)
print("Top 3 thấp nhất theo từng mã môn học:")
display(df_bottom3_course)

Top 3 thấp nhất theo từng mã môn học:


,student_id,name,class,course_id,score,ranking_in_course
0,1,Nguyen Minh Hoang,May Tinh,12,6.7,1
1,9,Duong Huu Phuc,Kinh Te,12,7.2,2
2,3,Pham Van Nam,Toan Tin,26,7.9,1
3,10,Cao Thi Hanh,May Tinh,34,7.0,1
4,2,Tran Thi Lan,Kinh Te,34,9.2,2
5,7,Bui Tien Dung,Kinh Te,34,9.2,2


Kết luận câu 3
+ a. Xếp hạng theo điểm số toàn bộ
* Khi xếp hạng toàn bộ theo điểm số, Tran Thi Lan và Bui Tien Dung cùng đạt điểm 9.2 nên đồng hạng 1. Sinh viên đứng tiếp theo là Pham Van Nam với điểm 7.9, hạng 3.
* Top 3 sinh viên điểm cao nhất toàn bộ gồm: Tran Thi Lan, Bui Tien Dung và Pham Van Nam.
* Top 3 sinh viên điểm thấp nhất toàn bộ gồm: Nguyen Minh Hoang, Cao Thi Hanh và Duong Huu Phuc
+ b. Xếp hạng theo từng lớp
* Ở lớp Kinh Te, Tran Thi Lan và Bui Tien Dung cùng đứng hạng 1, còn Duong Huu Phuc đứng hạng 3.
* Ở lớp May Tinh, Cao Thi Hanh đứng hạng 1, Nguyen Minh Hoang đứng hạng 2.
* Ở lớp Toan Tin, chỉ có Pham Van Nam, nên sinh viên này đứng hạng 1 trong lớp.
* Việc lấy top 3 cao nhất và thấp nhất theo từng lớp đã phản ánh đúng thứ hạng trong từng nhóm lớp học.
+ c. Xếp hạng theo từng mã môn học
* Với môn 12, Duong Huu Phuc đứng hạng 1, Nguyen Minh Hoang đứng hạng 2.
* Với môn 26, chỉ có Pham Van Nam, nên sinh viên này đứng hạng 1.
* Với môn 34, Tran Thi Lan và Bui Tien Dung cùng hạng 1, còn Cao Thi Hanh đứng hạng 3.
* Kết quả top 3 cao nhất và thấp nhất theo mã môn học đã cho thấy sự khác nhau về điểm số giữa các sinh viên trong từng môn.

In [ ]:
# Câu 4
# 1. Thêm cột graduation_date
cursor.execute("""
ALTER TABLE student
ADD COLUMN graduation_date DATETIME
""")
conn.commit()

print("Sau khi thêm cột graduation_date:")
display(pd.read_sql("SELECT * FROM student", conn))

Sau khi thêm cột graduation_date:


,student_id,name,class,course_id,score,graduation_date
0,1,Nguyen Minh Hoang,May Tinh,12,6.7,None
1,2,Tran Thi Lan,Kinh Te,34,9.2,None
2,3,Pham Van Nam,Toan Tin,26,7.9,None
3,7,Bui Tien Dung,Kinh Te,34,9.2,None
4,9,Duong Huu Phuc,Kinh Te,12,7.2,None
5,10,Cao Thi Hanh,May Tinh,34,7.0,None


In [ ]:
# 2. Xếp hạng sinh viên theo điểm số
query_rank = """
SELECT
    student_id,
    name,
    score,
    RANK() OVER (ORDER BY score DESC) AS ranking
FROM student
"""
df_rank = pd.read_sql(query_rank, conn)
print("Bảng xếp hạng sinh viên theo điểm số:")
display(df_rank)

Bảng xếp hạng sinh viên theo điểm số:


,student_id,name,score,ranking
0,2,Tran Thi Lan,9.2,1
1,7,Bui Tien Dung,9.2,1
2,3,Pham Van Nam,7.9,3
3,9,Duong Huu Phuc,7.2,4
4,10,Cao Thi Hanh,7.0,5
5,1,Nguyen Minh Hoang,6.7,6


In [ ]:
# 3. Cập nhật graduation_date theo hạng
cursor.execute("""
UPDATE student
SET graduation_date = (
    SELECT datetime('now', '+' || ranking || ' days')
    FROM (
        SELECT
            student_id,
            RANK() OVER (ORDER BY score DESC) AS ranking
        FROM student
    ) AS ranked
    WHERE ranked.student_id = student.student_id
)
""")
conn.commit()

In [ ]:
# 4. Hiển thị kết quả cuối cùng
query_result = """
SELECT
    student_id,
    name,
    class,
    course_id,
    score,
    graduation_date
FROM student
ORDER BY score DESC
"""
df_result = pd.read_sql(query_result, conn)
print("Bảng student sau khi cập nhật graduation_date:")
display(df_result)

Bảng student sau khi cập nhật graduation_date:


,student_id,name,class,course_id,score,graduation_date
0,2,Tran Thi Lan,Kinh Te,34,9.2,2026-03-29 14:19:21
1,7,Bui Tien Dung,Kinh Te,34,9.2,2026-03-29 14:19:21
2,3,Pham Van Nam,Toan Tin,26,7.9,2026-03-31 14:19:21
3,9,Duong Huu Phuc,Kinh Te,12,7.2,2026-04-01 14:19:21
4,10,Cao Thi Hanh,May Tinh,34,7.0,2026-04-02 14:19:21
5,1,Nguyen Minh Hoang,May Tinh,12,6.7,2026-04-03 14:19:21


In [ ]:
# 5. Có thể xem kèm thứ hạng để dễ giải thích
query_final = """
SELECT
    student_id,
    name,
    class,
    course_id,
    score,
    RANK() OVER (ORDER BY score DESC) AS ranking,
    graduation_date
FROM student
ORDER BY ranking, student_id
"""
df_final = pd.read_sql(query_final, conn)
print("Kết quả cuối cùng kèm thứ hạng:")
display(df_final)

Kết quả cuối cùng kèm thứ hạng:


,student_id,name,class,course_id,score,ranking,graduation_date
0,2,Tran Thi Lan,Kinh Te,34,9.2,1,2026-03-29 14:19:21
1,7,Bui Tien Dung,Kinh Te,34,9.2,1,2026-03-29 14:19:21
2,3,Pham Van Nam,Toan Tin,26,7.9,3,2026-03-31 14:19:21
3,9,Duong Huu Phuc,Kinh Te,12,7.2,4,2026-04-01 14:19:21
4,10,Cao Thi Hanh,May Tinh,34,7.0,5,2026-04-02 14:19:21
5,1,Nguyen Minh Hoang,May Tinh,12,6.7,6,2026-04-03 14:19:21


Kết luận câu 4
* Bảng student đã được bổ sung thêm trường graduation_date với kiểu dữ liệu DATETIME.
* Thời gian tốt nghiệp của mỗi sinh viên được xác định bằng thời điểm hiện tại cộng với số hạng tương ứng của sinh viên đó trong bảng xếp hạng theo điểm số.
* Hai sinh viên có điểm cao nhất là Tran Thi Lan và Bui Tien Dung cùng hạng 1, nên có cùng graduation_date sớm nhất.
* Sinh viên có thứ hạng thấp hơn sẽ có graduation_date muộn hơn theo đúng quy tắc đề bài đã yêu cầu.